File: URI/V1.2.8/python/src/iqa/attest.py
"""
attest -- the sovereign **attestation envelope** (Ed25519, self-certifying)

Why this module exists
----------------------
An ``iqa://`` URI is a *claim about standing* (RFC-009 sec. 10.4: "parsing is not
attestation"). To act on the claim, a verifier needs evidence. The evidence
travels as an **envelope**: a signed object that cites the subject URI and
carries the standing an Organ rendered.

Layout: identical to the RTTP sovereign seal
(``SPEC/RTTP-SEAL-ENVELOPE-v1.2.6.md`` sec. 3) with two changes fixed by
``SPEC/IQA-URI-ATTEST-v1.2.6.md`` sec. 5:

    1. **Domain separation.** The signing-input prefix is ``b"iqa-attest-v1\\n"``
       -- a signature made for one scheme can never be replayed as the other.
    2. **Payload semantics.** ``{"subject_uri", "organ", "standing"}``, with
       ``standing`` a closed set (``ghost`` / ``probation`` / ``radiant`` /
       ``genesis``, RFC-009-C sec. 3).

Identity: ``AID = SHA-256(public key)`` -- self-certifying, recomputed by the
verifier from the in-band key. No registry, no issuer, no directory. Lose the
key, lose the identity; there is no operator who can restore it.

``ts`` and ``nonce`` are **inside** the signature: rewriting them must require
breaking Ed25519, not just editing JSON. Freshness window: 120 seconds for any
envelope arriving over a network; archival validation only may disable it.

Installing
----------
    pip install iqa[ed25519]

Without the extra, importing this module succeeds but every call raises
``AttestError`` with that instruction -- it never degrades to an unsigned path.
"""

from __future__ import annotations

import hashlib
import json
import os
import time

from . import iqa_uri
from .iqa_uri import IqaUriError

try:
    from cryptography.exceptions import InvalidSignature
    from cryptography.hazmat.primitives import serialization
    from cryptography.hazmat.primitives.asymmetric.ed25519 import (
        Ed25519PrivateKey,
        Ed25519PublicKey,
    )

    HAVE_ED25519 = True
    _IMPORT_ERROR = None
except ImportError as _exc:  # pragma: no cover - depends on the environment
    HAVE_ED25519 = False
    _IMPORT_ERROR = _exc

    class InvalidSignature(Exception):  # type: ignore[no-redef]
        """Placeholder used only when no Ed25519 backend is installed.

        It is never raised: without a backend the guard raises ``AttestError``
        first. It exists so that ``except InvalidSignature:`` is always a
        resolvable name. Without it, an install that omits the ``[ed25519]``
        extra dies with ``NameError`` while evaluating the ``except`` clause,
        instead of failing closed with the documented error.
        """

        def __init__(self) -> None:  # pragma: no cover
            super().__init__("no Ed25519 backend is installed")

ALG = "ed25519"
ENVELOPE_VERSION = 1

#: Maximum accepted |now - ts|, in seconds.
MAX_CLOCK_SKEW = 120

#: Nonce length in bytes (hex-encoded on the wire).
NONCE_BYTES = 8

AID_BYTES = 32
PUBLIC_KEY_BYTES = 32
SIGNATURE_BYTES = 64
SEED_BYTES = 32

#: SPEC/IQA-URI-ATTEST-v1.2.6 sec. 5.2 -- domain separation from the RTTP seal.
CANONICAL_PREFIX = b"iqa-attest-v1\n"

#: Closed set of standings an Organ may render (RFC-009-C sec. 3).
STANDINGS = iqa_uri.STANDINGS

INSTALL_HINT = "pip install iqa-org[ed25519]"


class AttestError(Exception):
    """The envelope is unavailable, unparsable or malformed."""


def _require_backend() -> None:
    if not HAVE_ED25519:
        raise AttestError(
            f"Ed25519 backend unavailable - install it with '{INSTALL_HINT}' "
            f"(import error: {_IMPORT_ERROR})")


# ---------------------------------------------------------------------------
# Signature backend (injectable)
# ---------------------------------------------------------------------------
#
# A default install (zero dependencies) has no `cryptography`, yet the *rule*
# layer of an envelope - version, algorithm, hex shape, ts and payload types,
# freshness, self-certification - needs no primitive at all. The self-test
# therefore runs that layer always, and the flow layer (round-trip,
# determinism, tamper matrix) with an injected test signer, which is what proves
# the binding property: which fields the signed bytes cover.
#
# The test signer is NOT Ed25519. The real primitive is proven separately by the
# RFC 8032 vectors. Injection happens only through a function argument: there is
# no mutable module state and no environment switch, so a production call cannot
# be silently redirected to a weaker backend.

class AttestBackend:
    """Signature backend interface. Default implementation: Ed25519."""

    name = "abstract"

    def available(self) -> bool:
        return False

    def from_seed(self, seed: bytes):
        raise AttestError(f"{self.name} backend unavailable")

    def generate(self):
        raise AttestError(f"{self.name} backend unavailable")

    def private_bytes(self, private_key) -> bytes:
        raise AttestError(f"{self.name} backend unavailable")

    def public_bytes(self, private_key) -> bytes:
        raise AttestError(f"{self.name} backend unavailable")

    def sign(self, private_key, message: bytes) -> bytes:
        raise AttestError(f"{self.name} backend unavailable")

    def verify(self, public_key: bytes, signature: bytes, message: bytes) -> bool:
        raise AttestError(f"{self.name} backend unavailable")


class CryptographyBackend(AttestBackend):
    """Native backend: standard Ed25519 as provided by `cryptography`."""

    name = "cryptography"

    def available(self) -> bool:
        return HAVE_ED25519

    def from_seed(self, seed: bytes):
        return Ed25519PrivateKey.from_private_bytes(bytes(seed))

    def generate(self):
        return Ed25519PrivateKey.generate()

    def private_bytes(self, private_key) -> bytes:
        return private_key.private_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PrivateFormat.Raw,
            encryption_algorithm=serialization.NoEncryption(),
        )

    def public_bytes(self, private_key) -> bytes:
        return private_key.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )

    def sign(self, private_key, message: bytes) -> bytes:
        return private_key.sign(message)

    def verify(self, public_key: bytes, signature: bytes, message: bytes) -> bool:
        try:
            Ed25519PublicKey.from_public_bytes(public_key).verify(signature, message)
            return True
        except InvalidSignature:
            return False


NATIVE_BACKEND = CryptographyBackend()


def _resolve_backend(backend=None) -> AttestBackend:
    chosen = NATIVE_BACKEND if backend is None else backend
    if chosen.available():
        return chosen
    raise AttestError(
        f"{chosen.name} backend unavailable - install it with '{INSTALL_HINT}' "
        f"(import error: {_IMPORT_ERROR})")


# ---------------------------------------------------------------------------
# Identity
# ---------------------------------------------------------------------------

def aid_from_public_key(public_key: bytes) -> bytes:
    """AID = SHA-256(public_key), 32 bytes -- self-certifying by construction."""
    if not isinstance(public_key, (bytes, bytearray)):
        raise AttestError("public_key must be bytes")
    if len(public_key) != PUBLIC_KEY_BYTES:
        raise AttestError(f"public_key must be {PUBLIC_KEY_BYTES} bytes")
    return hashlib.sha256(bytes(public_key)).digest()


def generate_keypair(seed: bytes = None,
                     backend: AttestBackend = None) -> dict:
    """Generate an Ed25519 key pair and its AID.

    :param seed: 32-byte seed. **For reproducible test vectors only**; omit in
                 production and let the OS entropy source decide.
    :param backend: signature backend. Omitted = the native Ed25519 one; the
                 self-test injects a test signer so the flow layer can run
                 without `cryptography`.
    """
    chosen = _resolve_backend(backend)
    if seed is None:
        private_key = chosen.generate()
    else:
        if not isinstance(seed, (bytes, bytearray)) or len(seed) != SEED_BYTES:
            raise AttestError(f"seed must be {SEED_BYTES} bytes")
        private_key = chosen.from_seed(seed)

    raw_seed = chosen.private_bytes(private_key)
    public_key = chosen.public_bytes(private_key)
    aid_bytes = aid_from_public_key(public_key)

    return {
        "private_key": private_key,          # signing use -- never share
        "seed": raw_seed,                    # 32B seed -- never share
        "public_key": public_key,            # 32B
        "aid": aid_bytes,                    # 32B = sha256(public_key)
        "seed_hex": raw_seed.hex(),
        "public_hex": public_key.hex(),
        "aid_hex": aid_bytes.hex(),
        "backend": chosen,                   # kept so seal() uses the same one
    }


def public_bundle(keypair: dict) -> dict:
    """The safely publishable part (identity + public key), no private material."""
    return {"alg": ALG, "aid": keypair["aid_hex"], "pub": keypair["public_hex"]}


# ---------------------------------------------------------------------------
# Canonical signing input
# ---------------------------------------------------------------------------

def _hex_lower(value, what: str, nbytes: int) -> str:
    if not isinstance(value, str):
        raise AttestError(f"{what} must be a hex string")
    if value != value.lower():
        raise AttestError(f"{what} must be lowercase hex (RFC-009 sec. 10.3 discipline)")
    if len(value) != nbytes * 2:
        raise AttestError(f"{what} must be {nbytes} bytes ({nbytes * 2} hex chars)")
    try:
        bytes.fromhex(value)
    except ValueError as exc:
        raise AttestError(f"{what} is not valid hex: {exc}") from exc
    return value


def json_canonical(obj) -> bytes:
    """Deterministic JSON encoding (UTF-8, sorted keys, no insignificant whitespace)."""
    return json.dumps(obj, sort_keys=True, separators=(",", ":"),
                      ensure_ascii=False).encode("utf-8")


def signing_input(alg: str, aid_hex: str, pub_hex: str, ts: int, nonce: str,
                  payload: dict) -> bytes:
    """The signed byte sequence. Every envelope field except ``sig`` enters it.

    ``ts`` / ``nonce`` MUST be inside the signature -- beside it, rewriting the
    replay window would not touch Ed25519 at all.
    """
    return CANONICAL_PREFIX + json_canonical({
        "alg": alg,
        "aid": aid_hex,
        "pub": pub_hex,
        "ts": ts,
        "nonce": nonce,
        "payload": payload,
    })


# ---------------------------------------------------------------------------
# Attestation claims
# ---------------------------------------------------------------------------

def build_claim(subject_uri: str, organ: str, standing: str) -> dict:
    """Build and validate an attestation payload (closed sets enforced).

    The subject URI must parse; ``organ`` and ``standing`` must be members of
    their closed sets. A claim that violates the grammar is not built -- it is
    rejected, so an envelope can never carry a subject URI the codec itself
    would reject.
    """
    parsed = iqa_uri.parse(subject_uri)  # raises IqaUriError on any violation
    if organ not in iqa_uri.ORGANS:
        raise AttestError(f"organ {organ!r} is outside the closed set {list(iqa_uri.ORGANS)}")
    if standing not in STANDINGS:
        raise AttestError(f"standing {standing!r} is outside the closed set {list(STANDINGS)}")
    if organ != parsed["organ"]:
        raise AttestError("organ does not match the subject URI's organ")
    return {
        "subject_uri": parsed["canonical_uri"],
        "organ": organ,
        "standing": standing,
    }


# ---------------------------------------------------------------------------
# Seal / verify
# ---------------------------------------------------------------------------

def seal(payload: dict, keypair: dict, ts: int = None, nonce: str = None) -> dict:
    """Seal a payload into a self-certifying envelope with your own key.

    :param ts:    unix seconds. Omitted = now. **Pass it only to produce
                  reproducible vectors** -- in production it pins the replay window.
    :param nonce: ``NONCE_BYTES`` bytes as lowercase hex. Omitted = OS entropy.
    """
    if not isinstance(payload, dict):
        raise AttestError("payload must be a dict")
    if not isinstance(keypair, dict) or "private_key" not in keypair:
        raise AttestError("keypair must come from generate_keypair()")
    chosen = keypair.get("backend") or _resolve_backend()

    if ts is None:
        ts = int(time.time())
    if not isinstance(ts, int) or isinstance(ts, bool):
        raise AttestError("ts must be an integer (unix seconds)")

    if nonce is None:
        nonce = os.urandom(NONCE_BYTES).hex()
    _hex_lower(nonce, "nonce", NONCE_BYTES)

    aid_hex = keypair["aid_hex"]
    pub_hex = keypair["public_hex"]
    signature = chosen.sign(
        keypair["private_key"],
        signing_input(ALG, aid_hex, pub_hex, ts, nonce, payload))

    return {
        "v": ENVELOPE_VERSION,
        "alg": ALG,
        "aid": aid_hex,
        "pub": pub_hex,
        "ts": ts,
        "nonce": nonce,
        "sig": signature.hex(),
        "payload": payload,
    }


def check_envelope(envelope: dict, now: int = None,
                   max_skew: int = MAX_CLOCK_SKEW,
                   check_freshness: bool = True) -> tuple:
    """Envelope rules: everything except the signature.

    This layer needs no cryptographic primitive, so a default install (zero
    dependencies) runs all of it. Returns ``(ok, reason)``. Any failed step is a
    False -- "could not check" is never treated as "passed".

    :param check_freshness: default True. False is **only** for archived
        envelopes (audit records, test vectors). On live network input it must
        stay True -- turn it off and replay attacks become trivial.
    """
    if not isinstance(envelope, dict):
        return False, "envelope must be a dict"
    if envelope.get("v") != ENVELOPE_VERSION:
        return False, f"unsupported envelope version: {envelope.get('v')!r}"
    if envelope.get("alg") != ALG:
        return False, f"unsupported algorithm: {envelope.get('alg')!r}"
    try:
        aid_hex = _hex_lower(envelope.get("aid"), "aid", AID_BYTES)
        pub_hex = _hex_lower(envelope.get("pub"), "pub", PUBLIC_KEY_BYTES)
        _hex_lower(envelope.get("sig"), "sig", SIGNATURE_BYTES)
        _hex_lower(envelope.get("nonce"), "nonce", NONCE_BYTES)
    except AttestError as exc:
        return False, str(exc)

    ts = envelope.get("ts")
    if not isinstance(ts, int) or isinstance(ts, bool):
        return False, "ts must be an integer (unix seconds)"

    payload = envelope.get("payload")
    if not isinstance(payload, dict):
        return False, "payload must be an object"

    if check_freshness:
        reference = int(time.time()) if now is None else now
        drift = abs(reference - ts)
        if drift > max_skew:
            return False, (f"timestamp skew too large ({drift}s > {max_skew}s)"
                           " - replay?")

    # --- self-certification: identity is derived, not claimed ---
    if aid_from_public_key(bytes.fromhex(pub_hex)).hex() != aid_hex:
        return False, "AID does not match public key (self-certification failed)"
    return True, "rules ok"


def verify_envelope(envelope: dict, now: int = None,
                    max_skew: int = MAX_CLOCK_SKEW,
                    check_freshness: bool = True,
                    backend: AttestBackend = None) -> tuple:
    """Verify an envelope. Returns ``(ok: bool, reason: str, aid_hex | None)``.

    Any failed step returns False -- **a check that cannot be performed is a
    failure, never a pass.** Without an Ed25519 backend every call fails closed
    with the install hint; there is no unsigned path.

    :param check_freshness: default True. False is **only** for archived
        envelopes (audit records, test vectors). On live network input it must
        stay True -- turn it off and replay attacks become trivial.
    """
    ok, reason = check_envelope(envelope, now=now, max_skew=max_skew,
                                check_freshness=check_freshness)
    if not ok:
        return False, reason, None
    try:
        chosen = _resolve_backend(backend)

        aid_hex = envelope["aid"]
        pub_hex = envelope["pub"]
        sig_hex = envelope["sig"]
        nonce = envelope["nonce"]
        ts = envelope["ts"]
        payload = envelope["payload"]
        public_key = bytes.fromhex(pub_hex)

        if not chosen.verify(
                public_key, bytes.fromhex(sig_hex),
                signing_input(ALG, aid_hex, pub_hex, ts, nonce, payload)):
            return False, "bad seal (signature does not verify)", None
        return True, "sealed", aid_hex

    except AttestError as exc:
        return False, str(exc), None
    except (ValueError, TypeError, KeyError) as exc:
        return False, f"malformed envelope: {exc}", None
    except Exception as exc:  # noqa: BLE001 - fail closed on any backend error
        return False, f"bad seal ({type(exc).__name__}: {exc})", None


def verify_attestation(envelope: dict, expected_subject_uri: str = None,
                       now: int = None, check_freshness: bool = True,
                       backend: AttestBackend = None) -> tuple:
    """Verify an envelope **and** its claim payload (closed sets, URI grammar).

    The signature proves who said it; the claim checks make sure what they said
    is at least well-formed. Neither proves the standing is true -- trusting an
    attestation is a different act from verifying one.
    """
    ok, reason, _aid = verify_envelope(envelope, now=now,
                                       check_freshness=check_freshness,
                                       backend=backend)
    if not ok:
        return False, reason
    payload = envelope["payload"]
    try:
        claim = build_claim(payload.get("subject_uri"), payload.get("organ"),
                            payload.get("standing"))
    except (AttestError, IqaUriError) as exc:
        return False, f"claim payload malformed: {exc}", None
    if expected_subject_uri is not None:
        if claim["subject_uri"] != iqa_uri.parse(expected_subject_uri)["canonical_uri"]:
            return False, "subject URI mismatch (cited a different subject?)", None
    return True, "sealed", None


def unseal(envelope: dict, now: int = None, check_freshness: bool = True,
           backend: AttestBackend = None) -> dict:
    """Return the payload if the envelope verifies; raise AttestError otherwise."""
    ok, reason, _aid = verify_envelope(envelope, now=now,
                                       check_freshness=check_freshness,
                                       backend=backend)
    if not ok:
        raise AttestError(reason)
    return envelope["payload"]

File: URI/V1.2.8/rust/src/attest.rs
//! Sovereign attestation envelope -- SPEC/IQA-URI-ATTEST-v1.2.6 sec. 5 (Ed25519,
//! self-certifying, domain prefix `iqa-attest-v1`). **Feature-gated**
//! (`ed25519`): the default build stays zero-dependency and still replays the
//! envelope's signing input and AID self-certification.
//!
//! The envelope **carries a claim** that an Organ rendered a standing; it does
//! not **create** one. Verifying is not trusting (RFC-009 sec. 10.4).

use crate::iqa_uri;
use crate::json::Json;
use crate::sha256::{bytes_to_hex, hex_to_bytes, sha256};
use ed25519_dalek::{Signer, SigningKey, Verifier, VerifyingKey};
use std::fmt;
use std::time::{SystemTime, UNIX_EPOCH};

pub const DOMAIN: &[u8] = b"iqa-attest-v1\n";
pub const MAX_SKEW: u64 = 120;
pub const NONCE_BYTES: usize = 8;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AttestError(pub String);

impl fmt::Display for AttestError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}
impl std::error::Error for AttestError {}

pub struct Keypair {
    pub signing: SigningKey,
    pub pub_hex: String,
    pub aid_hex: String,
}

pub fn aid_from_public_key(public: &[u8; 32]) -> [u8; 32] {
    sha256(public)
}

pub fn generate_keypair() -> Result<Keypair, AttestError> {
    let mut seed = [0u8; 32];
    getrandom::getrandom(&mut seed).map_err(|e| AttestError(e.to_string()))?;
    Ok(keypair_from_seed(seed))
}

/// Deterministic keypair from a seed -- reproducible vectors only.
pub fn keypair_from_seed(seed: [u8; 32]) -> Keypair {
    let signing = SigningKey::from_bytes(&seed);
    let pub_bytes = signing.verifying_key().to_bytes();
    Keypair {
        signing,
        pub_hex: bytes_to_hex(&pub_bytes),
        aid_hex: bytes_to_hex(&aid_from_public_key(&pub_bytes)),
    }
}

/// Build and validate a claim payload (closed sets enforced). A claim that
/// violates the grammar is not built -- so an envelope can never carry a
/// subject URI the codec itself would reject.
pub fn build_claim(subject_uri: &str, organ: &str, standing: &str) -> Result<Json, AttestError> {
    let parsed = iqa_uri::parse(subject_uri).map_err(|e| AttestError(e.0))?;
    if !iqa_uri::ORGANS.contains(&organ) {
        return Err(AttestError(format!(
            "organ {:?} is outside the closed set [forge, tss, gateway]",
            organ
        )));
    }
    if !iqa_uri::STANDINGS.contains(&standing) {
        return Err(AttestError(format!("standing {:?} is outside the closed set", standing)));
    }
    if parsed.organ != organ {
        return Err(AttestError("organ does not match the subject URI's organ".into()));
    }
    Ok(Json::Obj(vec![
        ("subject_uri".into(), Json::Str(parsed.canonical_uri)),
        ("organ".into(), Json::Str(organ.to_string())),
        ("standing".into(), Json::Str(standing.to_string())),
    ]))
}

/// Canonical signing input: `iqa-attest-v1\n` + sorted-key JSON of every
/// envelope field except `sig`.
pub fn signing_input(env: &Json) -> Result<Vec<u8>, AttestError> {
    if !matches!(env, Json::Obj(_)) {
        return Err(AttestError("envelope must be an object".into()));
    }
    let inner = Json::Obj(vec![
        ("alg".into(), env.get("alg").cloned().unwrap_or(Json::Null)),
        ("aid".into(), env.get("aid").cloned().unwrap_or(Json::Null)),
        ("nonce".into(), env.get("nonce").cloned().unwrap_or(Json::Null)),
        ("payload".into(), env.get("payload").cloned().unwrap_or(Json::Null)),
        ("pub".into(), env.get("pub").cloned().unwrap_or(Json::Null)),
        ("ts".into(), env.get("ts").cloned().unwrap_or(Json::Null)),
    ]);
    let mut out = DOMAIN.to_vec();
    out.extend_from_slice(inner.canonical().as_bytes());
    Ok(out)
}

pub fn seal(
    payload: &Json,
    keypair: &Keypair,
    ts: Option<u64>,
    nonce_hex: Option<&str>,
) -> Result<Json, AttestError> {
    if !matches!(payload, Json::Obj(_)) {
        return Err(AttestError("payload must be an object".into()));
    }
    let ts = match ts {
        Some(t) => t,
        None => SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_err(|e| AttestError(e.to_string()))?
            .as_secs(),
    };
    let nonce = match nonce_hex {
        Some(n) => {
            if n.len() != NONCE_BYTES * 2
                || !n.bytes().all(|b| b.is_ascii_hexdigit())
                || n.bytes().any(|b| b.is_ascii_uppercase())
            {
                return Err(AttestError("nonce must be 8 bytes of lowercase hex".into()));
            }
            n.to_string()
        }
        None => {
            let mut raw = [0u8; NONCE_BYTES];
            getrandom::getrandom(&mut raw).map_err(|e| AttestError(e.to_string()))?;
            bytes_to_hex(&raw)
        }
    };
    let env = Json::Obj(vec![
        ("v".into(), Json::Num("1".into())),
        ("alg".into(), Json::Str("ed25519".into())),
        ("aid".into(), Json::Str(keypair.aid_hex.clone())),
        ("pub".into(), Json::Str(keypair.pub_hex.clone())),
        ("ts".into(), Json::Num(ts.to_string())),
        ("nonce".into(), Json::Str(nonce)),
        ("payload".into(), payload.clone()),
    ]);
    let sig = keypair.signing.sign(&signing_input(&env)?).to_bytes();
    if let Json::Obj(pairs) = &env {
        let mut out = pairs.clone();
        out.push(("sig".into(), Json::Str(bytes_to_hex(&sig))));
        return Ok(Json::Obj(out));
    }
    unreachable!()
}

/// Verify an envelope (fail closed, in order -- SPEC sec. 5). On success returns
/// the signer AID.
pub fn verify_envelope(
    env: &Json,
    now: Option<u64>,
    check_freshness: bool,
) -> Result<String, AttestError> {
    if env.get("v").and_then(Json::as_u128) != Some(1) {
        return Err(AttestError(format!("unsupported envelope version: {:?}", env.get("v"))));
    }
    if env.get("alg").and_then(Json::as_str) != Some("ed25519") {
        return Err(AttestError(format!("unsupported algorithm: {:?}", env.get("alg"))));
    }
    for (key, bytes) in [("aid", 32usize), ("pub", 32), ("sig", 64), ("nonce", 8)] {
        let v = env.get(key).and_then(Json::as_str).unwrap_or("");
        if v.len() != bytes * 2
            || !v.bytes().all(|b| b.is_ascii_hexdigit())
            || v.bytes().any(|b| b.is_ascii_uppercase())
        {
            return Err(AttestError(format!(
                "{} must be {} bytes of lowercase hex -- a case variant is rejected, never normalised",
                key, bytes
            )));
        }
    }
    let ts = env
        .get("ts")
        .and_then(Json::as_u128)
        .ok_or_else(|| AttestError("ts must be an integer (unix seconds)".into()))?;
    let ts = u64::try_from(ts).map_err(|_| AttestError("ts out of range".into()))?;
    if !matches!(env.get("payload"), Some(Json::Obj(_))) {
        return Err(AttestError("payload must be an object".into()));
    }
    if check_freshness {
        let now = now.unwrap_or_else(|| {
            SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_secs()).unwrap_or(0)
        });
        let drift = now.abs_diff(ts);
        if drift > MAX_SKEW {
            return Err(AttestError(format!(
                "timestamp skew too large ({}s > {}s) - replay?",
                drift, MAX_SKEW
            )));
        }
    }
    let pub_hex = env.get("pub").and_then(Json::as_str).unwrap_or("");
    let pub_bytes = hex_to_bytes(pub_hex).ok_or_else(|| AttestError("bad pub hex".into()))?;
    let mut pub32 = [0u8; 32];
    pub32.copy_from_slice(&pub_bytes);
    let aid = bytes_to_hex(&aid_from_public_key(&pub32));
    if aid != env.get("aid").and_then(Json::as_str).unwrap_or("") {
        return Err(AttestError("AID does not match public key (self-certification failed)".into()));
    }
    let key = VerifyingKey::from_bytes(&pub32).map_err(|e| AttestError(e.to_string()))?;
    let sig_bytes = hex_to_bytes(env.get("sig").and_then(Json::as_str).unwrap_or(""))
        .ok_or_else(|| AttestError("bad sig hex".into()))?;
    let mut sig64 = [0u8; 64];
    sig64.copy_from_slice(&sig_bytes);
    let sig = ed25519_dalek::Signature::from_bytes(&sig64);
    key.verify(&signing_input(env)?, &sig)
        .map_err(|_| AttestError("bad seal (signature does not verify)".into()))?;
    Ok(aid)
}

/// Verify an envelope **and** its claim payload (closed sets, URI grammar).
/// The signature proves who said it; the claim checks make sure what they said
/// is well-formed. Neither proves the standing is true.
pub fn verify_attestation(
    env: &Json,
    expected_subject_uri: Option<&str>,
    now: Option<u64>,
    check_freshness: bool,
) -> Result<(), AttestError> {
    verify_envelope(env, now, check_freshness)?;
    let payload = env.get("payload").ok_or_else(|| AttestError("payload missing".into()))?;
    let subject = payload.get("subject_uri").and_then(Json::as_str).unwrap_or("");
    let organ = payload.get("organ").and_then(Json::as_str).unwrap_or("");
    let standing = payload.get("standing").and_then(Json::as_str).unwrap_or("");
    let claim = build_claim(subject, organ, standing)?;
    if let Some(expected) = expected_subject_uri {
        let want = iqa_uri::parse(expected).map_err(|e| AttestError(e.0))?;
        let got = claim.get("subject_uri").and_then(Json::as_str).unwrap_or("");
        if got != want.canonical_uri {
            return Err(AttestError("subject URI mismatch (cited a different subject?)".into()));
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn seal_verify_roundtrip() {
        let kp = keypair_from_seed([7u8; 32]);
        let claim = build_claim("iqa://3f9a1b2c.gateway.iqa", "gateway", "radiant").unwrap();
        let env = seal(&claim, &kp, Some(1_760_000_000), Some("0011223344556677")).unwrap();
        assert!(verify_envelope(&env, Some(1_760_000_000), true).is_ok());
        assert!(verify_attestation(&env, Some("iqa://3f9a1b2c.gateway.iqa"),
                                   Some(1_760_000_000), true).is_ok());
        // Tamper: alter the standing after signing.
        if let Json::Obj(pairs) = &env {
            let mut t = pairs.clone();
            for (k, v) in t.iter_mut() {
                if k == "payload" {
                    *v = build_claim("iqa://3f9a1b2c.gateway.iqa", "gateway", "genesis").unwrap();
                }
            }
            assert!(verify_envelope(&Json::Obj(t), Some(1_760_000_000), false).is_err());
        }
    }

    #[test]
    fn claim_closed_sets_reject() {
        assert!(build_claim("iqa://3f9a1b2c.gateway.iqa", "gateway", "ascendant").is_err());
        assert!(build_claim("iqa://3f9a1b2c.forge.iqa", "gateway", "radiant").is_err());
        assert!(build_claim("iqa://3f9a1b2c.forgery.iqa", "forge", "radiant").is_err());
    }
}
